### ISSUE TITLE
PolymorphicDispatcher#compare violates the contract for Java Comparators

### ISSUE THREAD

**Issue by kduske-n4** [eclipse/xtext-core#238]
Sometimes, we get the following exception when our Xsemantics-generated type system is initialized:

```
1) Error injecting constructor, java.lang.IllegalArgumentException: Comparison method violates its general contract!
  at eu.numberfour.n4js.typesystem.CustomInternalTypeSystem.<init>(Unknown Source)
  at eu.numberfour.n4js.typesystem.CustomInternalTypeSystem.class(Unknown Source)
  while locating eu.numberfour.n4js.typesystem.CustomInternalTypeSystem
  while locating eu.numberfour.n4js.xsemantics.InternalTypeSystem
    for field at eu.numberfour.n4js.typesystem.N4JSTypeSystem.ts_internal(Unknown Source)
  at org.eclipse.xtext.service.MethodBasedModule.configure(MethodBasedModule.java:57)
  while locating eu.numberfour.n4js.typesystem.N4JSTypeSystem
    for field at eu.numberfour.n4js.validation.AbstractN4JSDeclarativeValidator.ts(Unknown Source)
  at org.eclipse.xtext.service.MethodBasedModule.configure(MethodBasedModule.java:57)
  while locating eu.numberfour.n4js.validation.validators.N4JSProjectSetupValidator
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.TimSort.mergeHi(TimSort.java:899)
	at java.util.TimSort.mergeAt(TimSort.java:516)
	at java.util.TimSort.mergeForceCollapse(TimSort.java:457)
	at java.util.TimSort.sort(TimSort.java:254)
	at java.util.Arrays.sort(Arrays.java:1512)
	at java.util.ArrayList.sort(ArrayList.java:1454)
	at java.util.Collections.sort(Collections.java:175)
	at org.eclipse.xtext.util.PolymorphicDispatcher.getDeclaredMethodsOrderedBySpecificParameterType(PolymorphicDispatcher.java:353)
	at org.eclipse.xtext.util.PolymorphicDispatcher.<init>(PolymorphicDispatcher.java:167)
	at org.eclipse.xtext.util.PolymorphicDispatcher.<init>(PolymorphicDispatcher.java:160)
	at it.xsemantics.runtime.XsemanticsRuntimeSystem$2.<init>(XsemanticsRuntimeSystem.java:86)
	at it.xsemantics.runtime.XsemanticsRuntimeSystem.buildPolymorphicDispatcher1(XsemanticsRuntimeSystem.java:86)
	at eu.numberfour.n4js.xsemantics.InternalTypeSystem.init(InternalTypeSystem.java:535)
	at eu.numberfour.n4js.xsemantics.InternalTypeSystem.<init>(InternalTypeSystem.java:531)
	at eu.numberfour.n4js.typesystem.CustomInternalTypeSystem.<init>(CustomInternalTypeSystem.java:39)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.google.inject.internal.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:85)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:85)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:65)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:54)
	at com.google.inject.internal.SingleFieldInjector.inject(SingleFieldInjector.java:53)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:110)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:94)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:65)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.SingleFieldInjector.inject(SingleFieldInjector.java:53)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:110)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:94)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:65)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:204)
	at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:198)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1024)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:198)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:179)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:109)
	at com.google.inject.Guice.createInjector(Guice.java:95)
	at com.google.inject.Guice.createInjector(Guice.java:72)
	at com.google.inject.Guice.createInjector(Guice.java:62)
	at eu.numberfour.n4js.N4JSInjectorProvider$1.createInjector(N4JSInjectorProvider.java:89)
	at eu.numberfour.n4js.N4JSStandaloneSetup.createInjectorAndDoEMFRegistration(N4JSStandaloneSetup.java:80)
	at eu.numberfour.n4js.N4JSInjectorProvider.internalCreateInjector(N4JSInjectorProvider.java:91)
	at eu.numberfour.n4js.N4JSInjectorProvider.getInjector(N4JSInjectorProvider.java:76)
	at eu.numberfour.n4js.N4JSInjectorProvider.setupRegistry(N4JSInjectorProvider.java:101)
	at org.eclipse.xtext.junit4.XtextRunner.methodBlock(XtextRunner.java:43)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:113)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:85)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:54)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:134)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

Upon further investigation, it turns out that Xtext's `PolymorphicDispatcher`, upon initialization, tries to sort all dispatch methods using a compare method that violates the general contract for Java Comparators. That contract specifies that a compare method should define a total order, i.e., it should be antisymmetric, transitive, and total.

However, there are pairs of methods `(a,b)` where the last requirement is violated. Such methods are incomparable to each other, that is, neither is `a <= b` nor `a <= b`. The compare method returns `0` for such cases, thereby indicating that `a == b`, which is obviously wrong. See the following example:

```
public class PolymorphicDispatcherTest extends PolymorphicDispatcher<Object> {

	public PolymorphicDispatcherTest() {
		super(Collections.singletonList(new Object()), method -> false);
	}

	private Method findMethod(Class<?> current, String name, Class<?>... parameterTypes) {
		if (current == Object.class)
			throw new IllegalArgumentException("Cannot find method");
		try {
			return current.getDeclaredMethod(name, parameterTypes);
		} catch (java.lang.NoSuchMethodException | SecurityException e) {
			return findMethod(current.getSuperclass(), name, parameterTypes);
		}
	}

        private static class S {}

	private static class A extends C {
	}

	private static class B extends S {
	}

	private static class C extends S {
	}

	public void test(A a) {
	}

	public void test(B b) {
	}

	public void test(C c) {
	}

	public void showTransitivityViolation() {
		MethodDesc m1 = createMethodDesc(this, findMethod(getClass(), "test", A.class));
		MethodDesc m2 = createMethodDesc(this, findMethod(getClass(), "test", B.class));
		MethodDesc m3 = createMethodDesc(this, findMethod(getClass(), "test", C.class));

		int m1m2 = PolymorphicDispatcherTest.this.compare(m1, m2); // 0
		int m2m3 = PolymorphicDispatcherTest.this.compare(m2, m3); // 0
		int m1m3 = PolymorphicDispatcherTest.this.compare(m1, m3); // 1

		if (m1m2 <= 0 && m2m3 <= 0 && !(m1m3 <= 0))
			throw new IllegalStateException("Transitivity violation");
	}

	public static void main(String[] args) {
		PolymorphicDispatcherTest test = new PolymorphicDispatcherTest();
		test.showTransitivityViolation();
	}
}
```

In this example, the method returns `0` when comparing method pairs `m1,m2` and `m2,m3`, even though in both cases, the methods are not identical! Comparing `m1` and `m3` then yields `1`, which violates transitivity.

We think the problem is due to the fact that the subtype relation, on which the compare method is based, only defines a partial order. The compare method attempts to extend this to a total order in an incorrect way.

This exception is thrown only sometimes, and it has popped up only recently (to our knowledge). We think that this is due to the fact that the initial order of the methods to be sorted depends on the initial order of the methods returned by `Class#getDeclaredMethods`, which is undefined by Java. Furthermore, Java uses different sorting algorithms depending on the number of elements to be sorted. In our particular cases, we are trying to sort 59 methods.

**Comment by JensN4** [comment]
To summarize here a very long discussion we (Kristian, Lorenzo and I) had, it seems to me that your bugfix, Moritz, is not working. The (initial) problem is that the Comparator used to compare methods is violating the constraints defined in the Java spec: it must define a total order -- and in this case it does not. Most sort methods and the SortedSet classes rely on that. It is pure luck if a sort call works when your Comparator does not define a total order. And it is also pure luck if your sort method sorts as expected when the Comparator is broken. 

In our case, the problem occurred in the context of Xsemantics, and Lorenzo created a first workaround (which did not work, but which seems to be similar to your "fix" -- which is why I'm pretty sure that your fix is not really a fix. Eventually we figured out that the dispatcher always traverses the whole list of methods, independent from the sorting. This explains why this bug does not lead to consequential problems in case the sort algorithm did not crash, but maybe produce wrong results. As a result we now use a second workaround provided by Lorenzo, in which the sorting is simply disabled -- and everything works fine now. It also doesn't seem to be a performance issue, since the result is cached anyway.

I tend to vote for this this bug to be one of the most serious bugs we found in the past years, as it effects the polymorphic dispatcher functionality which is one of the core things in Xtext (although in many cases Xtend dispatcher is used nowadays). But I wouldn't be surprised if this is a problem in XtendCompiler as well...

Just as a side node: We needed approximately 2 weeks of debugging due to this bug, since it is super hard to identify as it is not reproducible. Thus we only found it during our nightly builds, and in some cases it lead to completely weird consequential errors.

**Comment by meysholdt** [comment]
Hi Jens,

thank you for looking at the fix! I actually provided two of them. 

I assume you were trying my second fix b1cce46750ba7388f769429485369d7d8100bf77, which is indeed based on Lorenzo's workaround. I was under the assumption that the workaround was a valid workaround. Thank you for letting me know it isn't. 

My first fix 9566c3aacced48e25b49244d6ccc3e911e86b926 ports the XtendCompoliler's solution back to the `PolymorphicDispatcher`. That algorithm does provide a total order as it doesn't make use of `class1.isAssignableFrom(class2)`, which, as @kduske-n4 pointed out correctly, caused the comparison relation to be not transitive. 

While I like the idea of not sorting at all, I fear that it will lead to nondeterministic behavior. 

We haven't merged any of the fixes yet (I'm glad) because we were shortly before the 2.11. release, and the changes have a high risk of breaking existing semantics of code which worked (to my knowledge) just fine since 2011: f7ca755ca4dbdf9ce3023f7011dc6a0341008012.

**Comment by kduske-n4** [comment]
@meysholdt Two remarks.

1. If the logic of selecting a best matching method depends on the order of the pre-sorted method descriptors, it was always broken since the descriptors are sorted with a broken comparator. You can't rely on the result of using the comparator to sort anything because it's undefined.

2. I had a closer look at the [algorithm](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.util/src/org/eclipse/xtext/util/PolymorphicDispatcher.java#L254) that seaches for a "best matching" method when Lorenzo suggested not sorting the methods at all. These are my observations:

1. The algorithm always iterates over all method descriptors. It does not terminate early, so the fact that the descriptors are sorted does not influence when the algorithm terminates.
2. The algorithm inspects every invokable method in the list. Again, this filtering does not depend on the order.
3. It compares the currently inspected method to the current best candidate, which is the first item in the result list. If the current method is a better match according to compare, it replaces all candidates. Otherwise, it adds it to the candidates.

Since there's no early termination from the loop, all that might change depending on the sort order is the order of the method candidates in the result list. But if there is more than one methods, the dispatcher [calls](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.util/src/org/eclipse/xtext/util/PolymorphicDispatcher.java#L287) [handleAmbiguousMethods](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.util/src/org/eclipse/xtext/util/PolymorphicDispatcher.java#L312), which throws an exception.

To summarize: The order of the method descriptors only affects the order of the methods in the result list of method candidates. This only plays a role if there's more than one candidate, but in that case, an exception will be thrown.

With this in mind, I think it's safe to say that the sorting has no observable effect on polymorphic dispatch.

**Comment by kduske-n4** [comment]
Of course, just as I clicked on the "Comment" button, I found a caveat. The order of the methods in the result list may affect if the algorithm throws away the candidates found so far or not. The reason is that it [compares the currently inspected method only to the first method in the result list](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.util/src/org/eclipse/xtext/util/PolymorphicDispatcher.java#L262).

The methods are sorted in descending order of "generality", so to speak. And when the algorithm finds a "more special" method that still fits the method call, it will throw away the current candidates and replace them with the currently inspected method.

* If the methods were sorted correctly, then it will always inspect the better candidates after the "more general" candidates, and it will replace the "more general" candidates it has found so far with the better one.
* If the methods are not sorted at all, it may find the best candidate before finding other, more general, but matching candidates. In that case, these candidates get added to the result list, leading to an ambiguous method call.

So I have to correct myself; the algorithm does depend on the order of the methods. But I think this is fixable by looking at all of the results found so far instead of only looking at the first one, at a slight cost in performance.

**Comment by JensN4** [comment]
@meysholdt I would consider this bug extremely serious and I would even postpone the release due to it. The incorrect behavior of the Comparator at the moment has two consequences: 

1. Basically EVERY Xtext based system could break at any (random) time, since the sorting algorithm may detect the violated constraint and throw an exception.
2. Even if it doesn't break with an exception, I would assume that the sorting algorithm doesn't work as expected since one of its assumptions is violated. The means random behavior of the system.

I understand that it seems to be highly unlikely to happen: In our case, we kind of "needed" 59 polymorphic methods. But in principle this can already happen for much less methods. Nevertheless it breaks a core concept.

**Comment by LorenzoBettini** [comment]
@kduske-n4 I was about to comment about this but you came first ;)

without sorting, it could happen that the algorithm might not find an ambiguous situation. So, if there's no actual ambiguity it is OK, but if there's one, the algorithm might not find it (everytime). I don't think that it raises an ambiguity if it is not there.

**Comment by kduske-n4** [comment]
@LorenzoBettini I think it's the other way around, but that doesn't matter. Even with the suggested fix (compare currently inspected methods to all results instead of the first) will not work because that again would rely on the fact that `compare` defines a total order, which it doesn't.

In essence, the `compare` function is broken, and in consequence so is every code that uses it.

**Comment by JensN4** [comment]
@LorenzoBettini @meysholdt 

I think that the overall algorithm has to be changed. A comparator is simply not suitable for the problem: Its compare method indicates three states: less, equal, greater. What is needed are four states: less (i.e. less specific), equal (same signature), greater (i.e. more specific), incompatible. At the moment, equal and incompatible are merged into one state. This might work in case of Xtend since a validator can ensure that no two equal methods are defined -- but this is not possible for the dynamic case, is it? Using a distance method could be a solution. In that case, a distance of 0 would indicate equal methods, a negative or positive value less or more specific methods, and infinite (or max. integer) could indicate incompatible methods. But of course using a distance method is useless when used inside a compare method.

I highly recommend to get rid of the Comparator (in Xtext, but also in Xtend if it is used there), and use a different approach that can indicate the four cases. Since the result is cached anyway, it may be not that expensive to not use a pre-sorted collection. 

In
https://github.com/jpilgrim/mitra2/blob/master/plugins/de.jevopi.mitra2/src/de/jevopi/mitra2/engine/RuleResolver.java#L190
I have implemented an algorithm that does not rely on a pre-sported collection (which I think is the wrong data structure for that problem anyway), but which instead uses a distance (between types of arguments and parameters). In my algorithm negative values mean incompatible methods, and the higher the distance, the better the match (a highest distance is computed upfront to identify exact matches). I don't say that this algorithm is super efficient. But at least it doesn't come with an awkward, constraint-violating Comparator.

**Comment by LorenzoBettini** [comment]
Hi

Before I forget what I have in mind, I'll write my idea here, waiting for feedback.

Let's start from the assumption that we **can't sort method descriptions in advance**. Let's also assume that we are analysing only method descriptions that match the current arguments.

The current implementation of _compare_ tells you whether a method description is "better" than another one, the other way round ("worse") or simply unrelated.  I think these 3 states are enough to implement the polymorphic dispatcher (again, **let's forget about sorting**, and think about simply comparing two method descriptions).

Given the list of candidates m_i found so far with the invariant that it only contains "best unrelated" candidates; given a new matching candidate m, 

- if m is worse than at least one m_i then just don't consider m
- otherwise remove any existing worse m_i and add m as a candidate

these two single steps should keep the above invariant, shouldn't it?

In the end you either have an ambiguity (showing all the possible best candidates) or the single best candidate.

That should be enough. Or am I missing something?

**Comment by JensN4** [comment]
@LorenzoBettini Hmm... I really do not understand why to use a Comparator which is not suited. It may be possible to do it with a comparator, but it is at least very hard to understand. In other words: It's like if I would ask you whether you live north of Italy (-1), south of Italy (1), or not in Italy at all (0). If you can answer that question correctly (and truthfully), then you have convinced me of using the Comparator approach.

**Comment by LorenzoBettini** [comment]
I'm only saying that, once you have all the possible matching methods, you only need to reduce them to a list of "best unrelated", which was what the original polymorphic dispatcher used to do (of course, if the list is not a singleton you have an ambiguous call). To have such a list of "best unrelated" candidates, incrementally, you only need to tell whether one method is "better"/"worse" than another one or unrelated. And the current "compare" method already does that (independently from the fact that it was meant to be used as a Comparator).

**Comment by kduske-n4** [comment]
> you only need to tell whether one method is "better"/"worse" than another one or unrelated. 

@LorenzoBettini Unless I'm mistaken, I think you need to tell whether one method is better or worse than all others that are still under consideration (i.e. the current list of results). But then you are implicitly still relying on the fact that the compare function induces a total order because you have to find the current best match in the list of results. This essentially requires sorting the list of results, and then we're back in square one.

**Comment by LorenzoBettini** [comment]
@kduske-n4 No: if you start with a list of "best unrelated", then, at each iteration, to keep this invariant, you only must do the two operations I told before. I only need to compare two methods and I don't need a total order, that's the difference.

If the current method is worse than at least one "best unrelated" candidate then it's wrong to add it. Otherwise remove the existing candidates that are worse and add this method. The invariant still holds in both cases after this iteration.

The "all the others that are still under consideration" are already unrelated and "best up-to-now". So you don't need to tell whether the current method is better than all the others; you only need to discard it if it's worse than at least one of the current ones or to add it (after removing the ones that are worse, if any). Note that while doing this, the current method might not be comparable with some of the current list, but that's part of the invariant.

In the end you get a list; if it's not a singleton you get an ambiguity.

**Comment by kduske-n4** [comment]
@LorenzoBettini My brain is a bit wired by now, but a few comments:

> If the current method is worse than at least one "best unrelated" candidate then it's wrong to add it.

This assumes that the current method cannot be worse than one candidate and better than another  candidate at the same time. I think this assumption is correct.

Proof: Let `m` be the current method, let `c` and `d` be two candidates, and let `x < y` iff `compare(x, y) < 0`. Assume that `<` is transitive.

1. We have `!(c < d)` and `!(d < c)` because `c` and `d` are incomparable by the invariant.
2. WLOG, assume `c < m` and `m < d`. But then `c < d` by transitivity of `<`, which contradicts 1.
3. Therefore, the assumption `c < m` and `m < d` must be wrong.

> Otherwise remove the existing candidates that are worse

This also assumes that the compare function is transitive: If the current method is better than a best candidate, it is also better than all methods that are worse than that candidate.

> and add this method.

At this point, the method is unrelated to all candidates, so the invariant that all candidates are pairwise unrelated still holds.


tldr; If `compare` is transitive, this should work. My initial example falsely claimed that `compare` is not transitive, but as we discussed via email, it is transitivity of incomparability that is violated.
